Add fields to sort based on selector

garrensmith opened the pull request:
## Overview

This is an user experience improvement. If a selector has a sort field set
then an index is only valid if the prefix of the sort fields match the 
prefix of the index fields. 

e.g Index = [A, B, C] with Sort = [A, B] is a valid sort
but if Sort = [B, C] then it is not valid for this index.

If an indexed field in the selector is constant, eg {A: {$eq: 21}} then
we can add it to the sort list because it won't affect sorting and the
original sort will still be valid.

e.g Index = [A, B] with Sort = [B] and selector has {A: 1}. 
Then we can make the Sort = [A, B].
The sort will work as expected and this will increase the possibility
of the index being chosen. It also helps a user where they might not have
put the correct initial fields in the sort.

## Testing recommendations

See the tests in the PR for examples

## Checklist

- [x] Code is written and works correctly;
- [x] Changes are covered by tests;
- [ ] Documentation reflects the changes;

nickva commented on src/mango/src/mango_selector.erl:
```diff
@@ -638,11 +639,110 @@ has_required_fields_int([{[{Field, Cond}]} | Rest], RequiredFields) ->
+    false;
+
+is_constant_field_int([{[{<<"$and">>, Args}]}], Field) when is_list(Args) ->
+    lists:any(fun(Arg) -> is_constant_field_int(Arg, Field) end, Args);
```
lists:all instead of lists:any?

nickva replied:
Ah right I see the issue. Any makes sense then

davisp commented:
Also, it seems like normalize_sort_fields and all of the mayb_filter_sort_fields logic should actually be inside each index's implementation module. It seems weird to have special cased those module/index types outside that logic.

garrensmith commented:
thanks @davisp that was a really helpful review. I've done all the code changes you suggested. I've also moved all the `maybe_filter_sort_fields` into the index implementation modules. 

The implementation you suggested doesn't work for the case where
we might have a missing field in the middle of a sort list like [here](https://github.com/apache/couchdb/pull/1376/files#diff-cf1206e8c258c0a911b68eaaf80c8085R157)

I've also rewritten the explanation in the code. Hopefully that makes more sense. I've also edited the PR explanation above with the same text.

davisp commented:
@garrensmith I see the PR updated but nothing else. Did you forget to push your local changes?

tonysun83 commented on src/mango/src/mango_idx_view.erl:
```diff
@@ -511,3 +512,44 @@ range_pos(Low, Arg, High) ->
+
+
+maybe_filter_by_sort_fields(Idx, SortFields, Selector) ->
+    Cols = mango_idx:columns(Idx),
```
I noticed this call was moved from ```mango_idx``` into each of three separate index types. You don't use ref ```Idx``` anywhere else, so perhaps you could move this line back and just pass in ```Cols``` as in a parameter? Having said that, in the future it might be useful if we need the entire index record for other purposes. If that is the case, then we can leave it here.

garrensmith replied:
I was in two minds here. Initially, I only passed in the `Cols` but then I thought it might be better to support Idx in case we need it for another reason. I don't really mind. Having Idx might future proof it a little.

tonysun83 commented:
You have a great set of tests here. I'll try to think of a scenario tomorrow but was curious about our index selection when multiple indexes exist and how this change affects the index selection. I've seen scenarios where previously incorrect indexes become invalid and we switch to a non-performant index.  However, this seems like a situation where users can only improve their situation from no_usable_index or all_docs to having a usable index. I'll try to think of a scenario where we possibly don't want this to happen (doubt it but just want to be thorough).

davisp commented on src/mango/src/mango_idx.erl:
```diff
@@ -68,7 +68,7 @@ get_usable_indexes(Db, Selector, Opts) ->
     UsableIndexes1 = lists:filter(UsableFilter, UsableIndexes0),
 
-    case maybe_filter_by_sort_fields(UsableIndexes1, SortFields) of
+    case maybe_filter_by_sort_fields(UsableIndexes1, Selector, SortFields) of
```
My bad that I was vague on this earlier. What i was suggesting here was to move maybe_filter_by_sort_fields completely and just move that logic to the existing `is_usable` callback that already has all the information required. The thought was that `is_usable` shouldn't return true if the index can't satisfy the sort (cause then its not usable :).

davisp commented on src/mango/src/mango_idx_special.erl:
```diff
@@ -96,3 +98,8 @@ end_key([{_, _, '$lte', Key}]) ->
+
+
+maybe_filter_by_sort_fields(Idx, SortFields, _Selector) ->
+    Cols = mango_idx:columns(Idx),
```
Seems weird to be going back out to mango_idx from inside the callback module. Shouldn't this just be columns(Idx)?

davisp commented on src/mango/src/mango_idx_view.erl:
```diff
@@ -511,3 +512,44 @@ range_pos(Low, Arg, High) ->
+
+% This is an user experience improvement. If a selector has a sort field set
+% then an index is only valid if the prefix of the sort fields match the 
+% prefix of the index fields. 
```
This is still misleading. The index is only usable if SortFields is a prefix of the index columns. You have prefix of both which suggests columns `[A, B, C]` could support a sort over `[A, C]` since they have a common prefix of `[A]`.

davisp commented on src/mango/src/mango_idx_view.erl:
```diff
@@ -511,3 +512,44 @@ range_pos(Low, Arg, High) ->
+% The sort will work as expected and this will increase the possibility
+% of the index being choosen. It also helps a user where they might not have
+% put correct initial fields in the sort.
+normalize_sort_fields(Cols, SortFields, Selector) ->
```
Also still not liking the name of this function. There's nothing being normalized. Its about adding constants to the sort list. Though add_constant_fields also seems wrongish. Maybe `insert_constant_sorts` or something.

davisp commented on src/mango/test/18-json-sort.py:
```diff
@@ -0,0 +1,197 @@
+            }
+        }
+        self.db.find(selector, explain=True)
+        explain = self.db.find(selector, sort=["cars", "name"], explain=True)
```
I still can't convince myself that inserting in the middle of a sort is going to be correct 100% of the time. I can't convince myself it won't be correct either though. Just finding it a bit concerning I can't decide either way on this one.

davisp commented on src/mango/src/mango_idx_special.erl:
```diff
@@ -63,9 +65,11 @@ columns(#idx{def=all_docs}) ->
+is_usable(#idx{def=all_docs} = Idx, Selector, SortFields) ->
     Fields = mango_idx_view:indexable_fields(Selector),
-    lists:member(<<"_id">>, Fields).
+    lists:member(<<"_id">>, Fields) and maybe_filter_by_sort_fields(Idx, SortFields, Selector).
```
Can you rename `maybe_filter_by_sort_fields` `can_use_sort` like you did for views. Also no need to export it any longer.

davisp commented on src/mango/src/mango_idx.erl:
```diff
@@ -66,13 +66,12 @@ get_usable_indexes(Db, Selector, Opts) ->
+    case lists:filter(UsableFilter, UsableIndexes0) of
+        [] -> 
+            ?MANGO_ERROR({no_usable_index, missing_sort_index});
+        UsableIndexes1 -> 
```
No need to use the numbered suffix since we're just returning the list now.

davisp commented on src/mango/test/mango.py:
```diff
@@ -240,6 +240,7 @@ def find(self, selector, limit=25, skip=0, sort=None, fields=None,
         else:
             path = self.path("_find")
         r = self.sess.post(path, data=body)
+        print r.json()
```
Left over debug logging should be removed.

davisp commented on src/mango/src/mango_idx_special.erl:
```diff
@@ -96,3 +100,15 @@ end_key([{_, _, '$lte', Key}]) ->
     Key.
+
+
+% maybe_filter_by_sort_fields(Idx, SortFields, Selector) ->
```
Don't include commented out code.

tonysun83 commented:
I ran the text index tests and they all pass. So with text indexes, we have to sort with a data type. This means we won't conflict with text and json index sorts when fields are explicitly specified. However, one scenario that I tried out is if we have a index all text index and we do a sort such as:

```
{
  "selector": {"a":0, "b": {"$gt": 10}},
  "sort": ["b"]
}
```
In the old code, the index all index would be chosen.

In the new code, it's possible we choose a json index instead. I tried this out with one json index and one index all index. The json index gets chosen since it's given higher priority. Are we okay with this?

garrensmith commented:
@tonysun83 I think ideally we always prefer a json index over text or all docs. So that scenario makes sense.

davisp commented on src/mango/src/mango_idx_view.erl:
```diff
@@ -511,3 +512,45 @@ range_pos(Low, Arg, High) ->
+% Currently we only add constant fields that are prefixes to the sort fields
+% set by the user. We considered adding in constant fields after sort fields
+% but were not 100% sure that it would not affect the sorting of the query.
+add_constant_sort_fields(Cols, SortFields, Selector) ->
```
This function still seems overly complicated for what we're checking. Specifically how we're foldr'ing and relying on the subtlety that if a field isn't constant that its dropped which should theoretically guarantee that the prefix check will always fail. Also the fact that can_use_sort never checks directly and instead is applying the optimization logic rather than only checking if we can optimize once we find a need to.

Playing around a bit I came up with this recursive approach that I think at least seems easier to grok the behavior. 

```erlang
can_use_sort(_Cols, [], _Selector) ->
    true;
can_use_sort([], SortFields, Selector) ->
    false;
can_use_sort([Col | _] = Cols, [Col | _] = SortFields, _Selector) ->
    lists:prefix(SortFields, Cols);
can_use_sort([Col | RestCols], SortFields, Selector) ->
    case mango_selector:is_constant_field(Selector, Col) of
        true -> can_use_sort(RestCols, SortFields, Selector);
        false -> false
    end.
```

Which translates to a fairly straight forward reasoning:

1. If no sort, we can use this
2. If we run out of index columns we can't use this index
3. If current column is the start of the sort, return if sort is a prefix
4. If the current column is constant, drop it and continue, else return false

davisp commented:
Don't forget to squerge before merging!

tonysun83 commented on src/mango/src/mango_cursor.erl:
```diff
@@ -48,18 +48,12 @@
-    case length(UsableIndexes) of
-        0 ->
-            AllDocs = mango_idx:special(Db),
-            create_cursor(Db, AllDocs, Selector, Opts);
```
I'm probably missing something here but are we getting rid of the fallback to all_docs?

garrensmith replied:
No. The fallback happens in the mango_idx_special. The all_docs is added as an index. Previously what would happen is in the `maybe_filter_by_sort` if there were no sort fields defined it would return `{ok, []}`. That would then be returned to the above case statement which would then use the `mango_idx:special`. Now that check happens in the sort filtering inside the mango_idx_special.

tonysun83 replied:
ok, I see it now, thanks for clarifying!

tonysun83 commented:
+1
